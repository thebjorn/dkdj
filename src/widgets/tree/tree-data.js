

import dk from "../../dk-obj";
import Class from "../../lifecycle/coldboot/dk-class";
import counter from "../../core/counter";
import {json} from "../../browser/dk-client";

export class Leaf extends Class {
    constructor(...args) {
        super({
            id: null,
            kind: '',
            name: '',
            parent: null,
            path: [],
            root: undefined,
            children: [],
        }, ...args);
    }

    visit(location, ctx) {
        return ctx.leaf(location, this);
    }
}


export class Tree extends Leaf {
    visit(location, ctx) {
        return ctx.tree(location, this);
    }
}


export class Data extends Class {
    constructor(...args) {
        super({
            depth: 0,           // int
            height: 0,          // int
            roots: [],          // [id1,... idk]
            cache: {},          // id -> tree|leaf
        }, ...args);
        if (this.roots) this.map_objects(this);
    }
    
    map_objects({depth, height, roots, cache}) {
        if (Object.values(cache).every(v => v instanceof Tree || v instanceof Leaf)) return;
        
        this.depth = depth;
        this.height = height;
        this.roots = roots;
        this.cache = cache;
        
        const is_leaf = node => node.children.length === 0 && this.roots.indexOf(node.id) === -1;

        // convert all items in cache to Tree|Leaf objects
        Object.values(this.cache).forEach(node => {
            this.cache[node.id] = new (is_leaf(node) ? Leaf : Tree)(node);
        });

        // convert all child-ids to child objects.
        Object.values(this.cache).forEach(node => {
            node.children = node.children.map(id => this.cache[id]);
        });

        // convert all root-ids to objects.
        this.roots = this.roots.map(rootid => this.cache[rootid]);
    }
}


/*
 *  A Forrest is a tree structure with 1-n root nodes.
 *
 *  Each node _must_ have the following properties:
 *
 *    - name (the text to display in the tree)
 *    - children (a list of nodes that are this node's direct descendants).
 *
 *  a node _may_ in addition have
 *
 *    - kind (node type, a string, default will be set to 'tree' or 'leaf')
 *    - id (this should be a string to avoid surprises due to ordering issues
 *          in javascript, by default this will be set to an internally
 *          semi-random number).
 *
 */
export class Forrest {
    constructor(rootnodes) {
        this.depth = undefined;
        this.height = undefined;
        this.roots = null;
        this.cache = undefined;

        const self = this;
        this.cache = {};

        const iter = function iter(nodes, fn, parent) {
            if (!nodes) return;
            nodes.forEach(function (node) {
                fn(node, parent);
                iter(node._children, fn, node);
            });
        };

        // identify (id), categorize (kind), and linearize (cache)
        iter(rootnodes, function (node) {
            if (!node.id) node.id = counter('id');
            if (!node.kind) node.kind = node._children ? 'tree' : 'leaf';
            self.cache[node.id] = node;
        });

        // parent and path
        iter(rootnodes, function (node, parent) {
            node.parent = parent || null;
            if (node.parent) {
                node.path = node.parent.path.concat([node.parent.id]);
            } else {
                node.path = [];
            }
        });

        // childids
        iter(rootnodes, function (node) {
            node.children = (node._children || []).map(function (n) { return n.id; });
        });

        // root node
        rootnodes.forEach(function (r) {
            r.root = r.id;
            iter(r._children, function (node) {
                node.root = r.id;
            });
        });

        this.depth = Math.max.apply(null, this.cachevals().map(function (v) {
            return 1 + v.path.length;
        }));

        this.height = Math.max.apply(null, this.cachevals().map(function (v) {
            return v.children.length;
        }));
        this.height = Math.max(rootnodes.length, this.height);

        // this.outerHeight = this.cachevals().map(function (v) {
        //     return v.children.length + 1;
        // }).reduce(function (a, b) { return a+b;}, 0);

        this.roots = rootnodes;
    }

    cachekeys() {
        return Object.keys(this.cache);
    }

    cachevals() {
        const self = this;
        return this.cachekeys().map(function (key) {
            return self.cache[key];
        });
    }
}


export class DataSource extends Data {
    fetch() {
        dk.trigger(this, 'fetch-data-start', this);
        dk.trigger(this, 'fetch-data', this);
    }
}


export class JSonDataSource extends DataSource {
    constructor(...args) {
        super({
            cache: {},
            roots: [],
            depth: 0,
            height: 0,
        }, ...args);
        this.__fetched = false;
        this.fetch();
    }
    
    fetch() {
        if (this.__fetched) return;
        dk.trigger(this, 'fetch-data-start', this);
        this.map_objects(this);
        this.__fetched = true;
        dk.trigger(this, 'fetch-data', this);
        return this;
    }
}


export class AjaxDataSource extends JSonDataSource {
    constructor(...args) {
        super({
            url: '',
        }, ...args);
        this.__in_json = false;
    }
    
    /*
     *  Fetch data from url.
     */
    fetch() {
        if (this.__in_json) return;
        this.__in_json = true;

        const self = this;
        dk.trigger(this, 'fetch-data-start', this);
        return json({
            url: this.url,
            cache: true,
            success: function (data) {
                self.map_objects(data);
                self.__in_json = false;
                dk.trigger(self, 'fetch-data', self);
            }
        });
    }
}


const tree_data = {
    Leaf,
    Tree,
    Data,
    Forrest,
    DataSource,
    JSonDataSource,
    AjaxDataSource
};

export default tree_data;
