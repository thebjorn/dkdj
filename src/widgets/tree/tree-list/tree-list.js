
import dk from "../../../dk-obj";
import {UIWidget} from "../../../widgetcore/ui-widget";
import {dkwarning} from "../../../lifecycle/coldboot/dkwarning";


export class Leaf extends UIWidget {
    constructor(...args) {
        super({
            type: 'dk-tree-list-leaf',
            tree_data: null,     // dk.tree.Leaf
            structure: {
                classes: ['leaf']
            },
            template: {
                root: 'li'
            },
            
        }, ...args);
    }
    
    decorate_widget() {
        this.widget()
            .attr('nodeid', this.tree_data.id)
            .attr('kind', this.tree_data.kind)
            .addClass(this.tree_data.kind);
    }
    construct() {
        this.decorate_widget();
        this.widget().text(this.tree_data.name);
    }
    select() {
        this.trigger('select', this);
    }
    handlers() {
        const self = this;
        this.widget().on('click', function (e) {
            e.stopImmediatePropagation();
            self.select();
        });
    }
}

export class Tree extends UIWidget {
    constructor(...args) {
        super({
            tree_data: null,                 // dk.tree.Tree (could also be Leaf iff root w/o children).
            animation: undefined,       // jQuery animation to pass to .hide()/.show()
            children: null,
            collapsed: false,
            icons: {
                closed: 'plus-square-o:fw',
                open: 'minus-square-o:fw',
                childless: 'square-o:fw'
            },
            structure: {
                classes: ['tree'],
                icon: { template: '<dk-icon/>' },
                label: { template: '<span/>' },
                subtree: { template: '<ul/>' }
            },
            template: {root: 'li'},
            
        }, ...args);
    }
    
    /*
     *  Get currently correct icon.
     */
    get_icon() {
        if (this.tree_data.children.length === 0) return this.icons.childless;
        if (this.collapsed) return this.icons.closed;
        return this.icons.open;
    }
    construct() {
        this.widget()
            .attr('nodeid', this.tree_data.id)
            .attr('kind', this.tree_data.kind)
            .addClass(this.tree_data.kind);
        this.icon.attr('value', this.get_icon());
        this.label.text(this.tree_data.name);
    }
    collapse(boolval) {
        const subtree = this.widget('> .subtree');
        if (boolval === undefined) boolval = !this.collapsed;  // for toggle behavior
        if (boolval) {
            subtree.hide(this.animation);
            this.trigger('collapse', this);
        } else {
            subtree.show(this.animation);
            this.trigger('expand', this);
        }
        this.collapsed = boolval;
        this.icon.attr('value', this.get_icon());
    }
    handlers() {
        const self = this;
        this.widget('dk-icon, span').on('click', function (e) {
            self.collapse();
            e.stopImmediatePropagation();
        });
    }
}


/*
 *  dk.tree.list.TreeWidget is the outermost tree widget.
 *  It also implements the visitor protocol for dk.tree.Leaf|Tree.
 */
export class TreeWidget extends UIWidget {
    constructor(...args) {
        if (args.length > 0 && args[0].data && !args[0].tree_data) {
            dkwarning("TreeWidget created with .data and not .tree_data!");
        }
        super({
            type: 'dktree',
            tree_data: null,             // dk.tree.DataSource
            animation: undefined,   // to pass to .show()/.hide()
            initial_collapse: true, // should only roots (true) be showing?

            roots: [],
            nodes: {},
            treewidget: Tree,
            leafwidget: Leaf,

            structure: {
                classes: ['panel', 'panel-default'],
                panel_body: {
                    tree: {
                        template: '<ul/>',
                        css: {
                            paddingLeft: 0
                        }
                    }
                }
            },
        }, ...args);
        this.nodes = {};
    }

    construct() {
        dk.on(this.tree_data, 'fetch-data-start', () => this.start_busy());
        dk.on(this.tree_data, 'fetch-data', () => this.draw());
        dk.on(this.tree_data, 'fetch-data', () => this.end_busy());
    }

    draw(data) {
        if (!data && this.tree_data && this.tree_data.__fetched) data = this.tree_data;
        if (!data) {
            this.tree_data.fetch();
        } else {
            this.trigger('draw-start', this);
            dk.debug("start tree draw", new Date());
            this.panel_body.tree.empty();
            this.roots = this.tree_data.roots.map(r => {
                return r.visit(this.panel_body.tree, this);
            });
            if (this.initial_collapse) {
                this.roots.forEach(function (r) {
                    r.collapse(true);
                });
            }
            dk.debug("finished tree draw", new Date());
            this.trigger('draw-end', this);
        }
    }

    // visitor methods..
    leaf(location, node) {
        const self = this;
        const leaf = this.leafwidget.append_to(location, { tree_data: node, tree: self });
        this.nodes[node.id] = leaf;
        return leaf;
    }
    tree(location, node) {
        const self = this;
        const t = this.treewidget.append_to(location, {
            tree_data: node,
            tree: self,
            animation: this.animation
        });

        t.children = node.children.map(function (item) {
            const child = item.visit(t.subtree, self);
            child.parent = t;
            return child;
        });
        this.nodes[node.id] = t;
        return t;
    }
}
