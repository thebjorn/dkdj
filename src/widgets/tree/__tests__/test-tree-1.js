
import $ from 'jquery';
import page from "../../../widgetcore/dk-page";
import {TreeWidget} from "../tree-list/tree-list";
import {JSonDataSource} from "../tree-data";
import utidy from "../../../browser/dk-html";

/* 
 *  An empty tree should produce a panel with an empty `ul` tag.
 */
test("test-empty-tree-list", () => {
    document.body.innerHTML = `
        <div id="work"></div>
    `;
    const work = $('#work');
    page.initialize(document);

    TreeWidget.create_inside(work, {
        data: JSonDataSource.create({
            data: {}
        })
    });
    
    expect(work.html()).toMatchSnapshot();
    // expect(utidy(work.html())).toEqual(utidy(`
    //     <div class="dktree panel panel-default" id="tree-widget">
    //         <div class="panel-body" id="dk-bx">
    //             <ul style="padding-left: 0px;" class="tree" id="dk-bx"></ul>
    //         </div>
    //     </div>
    // `));
});


