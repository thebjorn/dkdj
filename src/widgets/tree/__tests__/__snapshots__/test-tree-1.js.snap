// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-empty-tree-list 1`] = `"<div id=\\"tree-widget-1\\" class=\\"dktree panel panel-default\\"><div id=\\"dk-bx-1\\" class=\\"panel-body\\"><ul id=\\"dk-bx-2\\" class=\\"tree\\" style=\\"padding-left: 0px;\\"></ul></div></div>"`;
